OpenRA.Mods.Cnc/Traits/World/PavementRenderer.cs

```diff
@@ -0,0 +1,206 @@
+	public class BuildableAutoTerrainRenderer : IRenderOverlay, IWorldLoaded, ITickRender, IRadarTerrainLayer, INotifyActorDisposing
+	{
+		readonly World world;
+		readonly PavementRendererInfo info;
+		readonly PavementLayer buildableAutoTerrainLayer;
+		readonly Dictionary<CPos, Sprite> dirty;
+		readonly ITiledTerrainRenderer terrainRenderer;
+		readonly CellLayer<(Color, Color)> radarColor;
+
+		TerrainSpriteLayer terrainSpriteLayer;
+		ITemplatedTerrainInfo terrainInfo;
+		PaletteReference paletteReference;
+
+		public BuildableAutoTerrainRenderer(Actor self, PavementRendererInfo info)
+		{
+			this.info = info;
+			world = self.World;
+			dirty = new Dictionary<CPos, Sprite>();
+			terrainRenderer = self.Trait<ITiledTerrainRenderer>();
+			buildableAutoTerrainLayer = self.Trait<PavementLayer>();
+			radarColor = new CellLayer<(Color, Color)>(world.Map);
+		}
+
+		void IWorldLoaded.WorldLoaded(World w, WorldRenderer wr)
+		{
+			paletteReference = wr.Palette(info.Palette);
+
+			terrainSpriteLayer = new TerrainSpriteLayer(w, wr, terrainRenderer.MissingTile, BlendMode.Alpha, wr.World.Type != WorldType.Editor);
+
+			if (!(w.Map.Rules.TerrainInfo is ITemplatedTerrainInfo terrainInfo))
+				throw new InvalidDataException(nameof(BuildableAutoTerrainRenderer) + " requires a template-based tileset.");
+
+			this.terrainInfo = terrainInfo;
+
+			buildableAutoTerrainLayer.Occupied.CellEntryChanged += Add;
+		}
+
+		[Flags]
+		enum ClearSides : byte
+		{
+			None = 0x0,
+			Left = 0x1,
+			Top = 0x2,
+			Right = 0x4,
+			Bottom = 0x8,
+
+			All = 0xFF
+		}
+
+		static readonly Dictionary<ClearSides, ushort> SpriteMap = new Dictionary<ClearSides, ushort>()
+		{
+			{ ClearSides.None, 671 },
+			{ ClearSides.Right, 597 },
+			{ ClearSides.Bottom, 598 },
+			{ ClearSides.Right | ClearSides.Bottom, 599 },
+			{ ClearSides.Left, 600 },
+			{ ClearSides.Left | ClearSides.Right, 601 },
+			{ ClearSides.Left | ClearSides.Bottom, 602 },
+			{ ClearSides.Left | ClearSides.Right | ClearSides.Bottom, 603 },
+			{ ClearSides.Top, 604 },
+			{ ClearSides.Top | ClearSides.Right, 605 },
+			{ ClearSides.Top | ClearSides.Bottom, 606 },
+			{ ClearSides.Top | ClearSides.Right | ClearSides.Bottom, 607 },
+			{ ClearSides.Top | ClearSides.Left, 608 },
+			{ ClearSides.Top | ClearSides.Right | ClearSides.Left, 609 },
+			{ ClearSides.Top | ClearSides.Left | ClearSides.Bottom, 609 },
+			{ ClearSides.All, 611 },
+		};
+
+		ClearSides FindClearSides(CPos cell)
+		{
+			var clearSides = ClearSides.None;
+			if (!buildableAutoTerrainLayer.Occupied[cell + new CVec(0, -1)])
+				clearSides |= ClearSides.Right;
+
+			if (!buildableAutoTerrainLayer.Occupied[cell + new CVec(-1, 0)])
+				clearSides |= ClearSides.Top;
+
+			if (!buildableAutoTerrainLayer.Occupied[cell + new CVec(1, 0)])
+				clearSides |= ClearSides.Bottom;
+
+			if (!buildableAutoTerrainLayer.Occupied[cell + new CVec(0, 1)])
+				clearSides |= ClearSides.Left;
+
+			return clearSides;
+		}
+
+		void Add(CPos cell)
```

Comment:
```suggestion
		void OnCellEntryChanged(CPos cell)
```
(or at least `OnCellUpdated`)

Reply:
No, there is no removal.